### DISCUSSION TITLE
I2C allowed pin assignments have to be paired?

### DISCUSSION THREAD

**Discussion by krismc** [micropython/micropython#16390] [Hardware & Peripherals]
I've found I can move uart Rx/Tx pins independently
![image](https://github.com/user-attachments/assets/b6706427-a566-4ad0-872f-58f0f4b18dcc)
but if I try to do likewise with i2c I get problems. Just want to confirm that they have to be moved in pairs & that combos like I2C0 scl on gp13 with sda on gp40 are not allowed?

**Comment by andrewleech** [comment]
Can you confirm which chip & micropython version you're using?
From the image it looks like esp32-s2 ?

**Reply by krismc** [reply]
Its a waveshare s2 running release 1.23.0

**Comment by robert-hh** [comment]
I never had the impression that I2C pins have to be moved in pairs. Maybe it's just the combination of gp13 and gp40, being in different 32bit blocks. 
N.b. interesting board with a pin layout like the rpi pico. Which brand is it?

**Reply by krismc** [reply]
Yeah I was a bit surprised I can't mix & match, but setup (with no i2c device) doesn't go well when I do....
```
from machine import SoftI2C, Pin; from time import time, sleep; import sys; t0=time()

i2c=SoftI2C(scl=Pin(41), sda=Pin(40), freq=10000); addr=i2c.scan(); print(time()-t0, 'i2c device addresses', addr)                                                          
if addr:
  i2c.writeto(40, b'')
  data=i2c.readfrom(40, 4)
  print(data)                                                                                            # from addr 40h read 4 bytes
else: print(time()-t0, 'no device'); t0=time()

i2c=SoftI2C(scl=Pin(13), sda=Pin(40), freq=10000); addr=i2c.scan(); print(time()-t0, 'i2c device addresses', addr)                                                          
if addr:
  i2c.writeto(40, b'')
  data=i2c.readfrom(40, 4)
  print(data)                                                                                  
else: print(time()-t0, 'no device'); t0=time()

i2c=SoftI2C(scl=Pin(13), sda=Pin(39), freq=10000); addr=i2c.scan(); print(time()-t0, 'i2c device addresses', addr)
if addr:
  i2c.writeto(40, b'')                                                                
  data=i2c.readfrom(40, 4)
  print(data)
else: print('no device')

0 i2c device addresses []
0 no device
11 i2c device addresses []
11 no device
11 i2c device addresses [8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119]
Traceback (most recent call last):
  File "<stdin>", line 21, in <module>
OSError: [Errno 19] ENODEV
```

**Reply by andrewleech** [reply]
That's even more strange; SoftI2C isn't using a hw peripheral so it really shouldn't matter which pins are used. Did you add the I2C pullups to the pins you're trying to use?

**Reply by krismc** [reply]
I didn't, I thought the way i2c works is that the assigned pins automatically become inputs with strong pullups when they're not outputs? It's been working with other i2c based pressure sensors.

**Comment by fdufnews** [comment]
I2C works with pullups on both lines.
[I2C bus specification](https://community.nxp.com/pwmxy87654/attachments/pwmxy87654/Power-Management/1156/1/UM10204.pdf)

**Comment by krismc** [comment]
I've got to the point where the the i2c device is at least found but I'm not making much progress in communicating with it using these instructions;
1. Send instructions 0x0A(combined conversion mode command) to 0x30 register to start conversion.
 2. Read the 0x30 register address. If Sco bit is 0, signify the acquisition end, the data can be read.
 (Or, skip above step and delay 20mS and to read data directly）
3. Read 0x06, 0x07, 0x08 register address data to form a 24-bit AD value (pressure data AD value); Read 0x09, 0x0A register
 address data to form a 16-bit AD value (temperature data AD value)
```
from machine import SoftI2C, Pin; from time import time, sleep; import sys; t0=time()
pwr=Pin(37,Pin.OUT,value=1,hold=0); sleep(.5)

i2c=SoftI2C(scl=Pin(12), sda=Pin(38), freq=10000); addr=i2c.scan(); print(time()-t0, 'i2c device addresses', addr)                                                          
if addr:
  i2c.writeto(0x6d, b'30')                                                                               # register 30h
  i2c.writeto(0x6d, b'0a')                                                                               # combined conversion
  sleep(.02)
  data=i2c.readfrom((0x30 & 0x08), 5)                                                         # from reg 30h read 5 bytes starting at 08h
  print(data)                                                                                            
else: print(time()-t0, 'no device'); t0=time()



0 i2c device addresses [109]
Traceback (most recent call last):
  File "<stdin>", line 11, in <module>
OSError: [Errno 19] ENODEV
```
Any suggestions on where I'm going wrong?

**Comment by krismc** [comment]
I think I've got comms now
```
if addr:
  i2c.writeto(0x6d, b'30')                                                                             # register 30h
  i2c.writeto(0x6d, b'0a')                                                                             # combined conversion
  sleep(.02)                                                                                           # wait for conversion to finish
  i2c.writeto(0x6d, b'30'); data=i2c.readfrom(0x6d, 90)                                                # reg 30h read 
  print(repr(data))                                                                                            
else: print(time()-t0, 'no device'); t0=time()

1 i2c device addresses [109]
b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00B\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```
but nothing that looks like pressure & temperature data?

**Comment by krismc** [comment]
Following

 _Read Temperature
 The bits of temperature output values in the 0x09 and 0x0A registers are 16 bits, the highest is the sign bit.
 The sign digit value is "0" when it represents "positive" temperature, and the sign digit value is "1" when it represents
 "negative" temperature.
 Supposing if the decimal values of REG0x09 and REG0x0A readout are X, Y,
 For Temperature ADC value and conversion formula are as:
 Temperature AD value: N=X*256+Y
 If n<2^15, Temperature is positive value, actual temperature T =N/256; (℃).
 If n>2^15, Temperature is negative value, actual temperature value = (N-2^16) /256; (℃)_
 
I get a semi plausible 32.1 celsius with
```
if addr:
  i2c.writeto_mem(0x6D, 30, b'\x0A')                                                                   # combined conversion
  sleep(.02)                                                                                           # wait for conversion to finish
  data=i2c.readfrom_mem(0x6D, 9, 2); print(repr(data), data[0], data[1])                                                
else: print(time()-t0, 'no device'); t0=time()

1 i2c device addresses [109]
b' \x12' 32 18
```
but it doesn't change when I spray the sensor with freeze-it spray so I'm still lost in the i2c wilderness. Jeez I find this stuff hard!
